### PR TITLE
feat: Add DCP All-to-All kernel for context-parallel attention reduction

### PR DESCRIPTION
**Pull request by davidjpyu**
## What is DCP?

**DCP (Decode Context Parallelism)** splits the KV cache across multiple GPUs for long-context inference. During the decode phase, each GPU computes partial attention over its local KV shard, producing two tensors:

- `partial_o` (bf16/fp16): partial attention output, shape `[B, cp_size, head_dim]`
- `softmax_stats` (fp32): softmax statistics for online correction, shape `[B, cp_size, stats_dim]`

These must be exchanged across all ranks (all-to-all) before the final attention reduction.

## Why this kernel?

The current approach uses **2× NCCL `all_to_all_single`** — one call per tensor since they have different dtypes. For the small message sizes in DCP (KB-range), NCCL protocol overhead dominates latency.

This PR adds a **fused LL128 FIFO kernel** (ported from TensorRT-LLM) that exchanges both tensors in a single kernel launch via MNNVL cross-GPU direct memory writes, eliminating NCCL protocol overhead.

## Kernel-level microbenchmark

GB200 NVL72, 4 GPUs, cp_size=4. Measures raw kernel time only (not end-to-end pipeline).

| batch | NCCL 2×a2a p50 | Native p50 | Speedup |
|-------|----------------|------------|---------|
| 1     | 0.176ms        | 0.035ms    | 5.0x    |
| 16    | 0.145ms        | 0.034ms    | 4.3x    |
| 64    | 0.140ms        | 0.033ms    | 4.3x    |
| 128   | 0.138ms        | 0.032ms    | 4.3x    |

Speedup is latency-dominated — these tensors are small, so NCCL protocol overhead is the bottleneck.

## Components

- **Kernel:** `csrc/nv_internal/tensorrt_llm/kernels/helixAllToAll.{cu,h}` — LL128 FIFO protocol, SM90+
- **Launcher + TVM-FFI binding:** `csrc/trtllm_dcp_alltoall.cu`
- **JIT generator:** `flashinfer/jit/comm.py` (`gen_dcp_alltoall_module`)
- **Python API:** `flashinfer/comm/dcp_alltoall.py`
  - `decode_cp_a2a_workspace_size` — query workspace bytes per rank
  - `decode_cp_a2a_allocate_workspace` — allocate workspace (MNNVL or device memory)
  - `decode_cp_a2a_init_workspace` — initialize FIFO buffers
  - `decode_cp_a2a_alltoall` — run all-to-all exchange
- **Tests:**
  - `tests/comm/test_dcp_alltoall.py` — 30 single-GPU tests (correctness, edge cases, input validation)
  - `tests/comm/test_mnnvl_dcp_alltoall.py` — 9 multi-GPU MNNVL tests (workspace shape, cross-rank visibility, alltoall correctness, FIFO reuse)
- **Benchmark:** `benchmarks/bench_dcp_alltoall.py` — native LL128 vs NCCL baseline
- **AOT registration** in `flashinfer/aot.py`

## Requirements

- SM90+ GPU (Hopper/Blackwell)
- MNNVL support for multi-GPU operation (GB200 NVL72 or similar)
- `mpi4py` for multi-GPU tests and benchmark

## Test plan

- [x] 30 single-GPU tests pass (`pytest tests/comm/test_dcp_alltoall.py`)
- [x] 9 multi-GPU MNNVL tests pass on 4× GB200 (`mpirun -np 4 pytest tests/comm/test_mnnvl_dcp_alltoall.py`)
- [x] Benchmark shows 4.3-5.0× speedup vs NCCL on GB200 NVL72
- [ ] CI passes (single-GPU tests; MNNVL tests auto-skip without fabric memory)

## Summary by CodeRabbit

* **New Features**
  * DCP decode-context-parallel all-to-all: GPU-native workspace allocation, init, and all-to-all invocation for faster multi-GPU attention reductions.
  * New Helix-based GPU all-to-all kernel plus low-level async-copy/barrier primitives to improve multi-GPU transfer performance.
  * Env-driven toggle to enable a new GPU launch mode on supported architectures.

* **Tests**
  * Added GPU microbenchmark and extensive GPU pytest suites validating correctness, workspace lifecycle, reuse, and MNNVL-backed multi-node integration.

* **Chores**
  * Communication modules and JIT build specs updated to target SM90+.

### CODE REVIEW

**Comment by davidjpyu** [comment]
**Re: aot.py — DCP module only generated for SM100+ (CodeRabbit)**

Fixed in 9f34778 — changed `if has_sm100:` to `if has_sm90 or has_sm100:`.

This also correctly covers `gen_trtllm_comm_module()`, `gen_trtllm_mnnvl_comm_module()`, and `gen_moe_alltoall_module()` in the same block, which all declare `supported_major_versions=[9, 10]`.

Verified all 30 DCP unit tests pass on H200 (SM90).

**Comment by davidjpyu** [comment]
**Re: Use `flashinfer.utils.get_compute_capability()` for SM90+ check (CodeRabbit)**

Adopted in a1e00ca — replaced custom `_sm90_available()` with `flashinfer.utils.get_compute_capability()`, matching the pattern in `tests/mamba/test_selective_state_update_stp.py`.

Also fixed the collection-time crash: module-level MNNVL allocation is now guarded behind `mnnvl_available() and _mpi4py_available()` checks, so CI environments without SYS_PTRACE or mpi4py will gracefully skip instead of erroring during collection.

File: csrc/nv_internal/cpp/common/envUtils.cpp [hunk 1]
```diff
@@ -356,4 +356,23 @@ int getEnvMoeA2ACombineBlockSize() {
 
 bool getEnvEplbForceGdrcopy() { return getBoolEnv("TRTLLM_EPLB_FORCE_GDRCOPY"); }
 
+bool getEnvEnablePDL() {
```
Comment (aleozlx) [inline csrc/nv_internal/cpp/common/envUtils.cpp]: we have been passing this as api argument (could be enable_pdl: bool = True)

Comment (davidjpyu) [reply csrc/nv_internal/cpp/common/envUtils.cpp]: It is now deleted in the forth commit and I've made it api argument instead

File: tests/comm/test_mnnvl_dcp_alltoall.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+
+
+# ─── Module-level MNNVL workspace (allocated once, reused across tests) ──
+
+# Guard module-level allocation: pytestmark skipif conditions are not
+# enforced during module import (collection phase). If we allocate
+# unconditionally, CI environments without SYS_PTRACE will fail at
+# collection time instead of gracefully skipping.
+if _sm90_available() and mnnvl_available() and _mpi4py_available():
+    _rank, _cp_size, _comm = _setup_rank()
```
[the remark this answers is not in the document]
Comment (davidjpyu) [reply tests/comm/test_mnnvl_dcp_alltoall.py]: without mpirun this test will be skipped.

File: csrc/nv_internal/tensorrt_llm/kernels/helixAllToAll.cu [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+
+  int preferredChannel = smCount / blockCountPerChannel;
+  return std::max(preferredChannel, 1);  // at least one channel
+}
+
+size_t computeHelixWorkspaceSizePerRank(int cpSize) {
+  static int maxChannelCount = 0;
+  if (maxChannelCount == 0) {
+    maxChannelCount = computeHelixMaxChannelCount(cpSize);
+  }
```
[the remark this answers is not in the document]
Comment (davidjpyu) [reply csrc/nv_internal/tensorrt_llm/kernels/helixAllToAll.cu]: Despite that cpSize will not change during a run, it is light weighted division so I fixed it now: maxChannelCount is no longer static and refreshed every time.

File: tests/comm/test_mnnvl_dcp_alltoall.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        assert _mnnvl_workspace.shape[0] == _cp_size, (
+            f"Expected workspace.shape[0] == {_cp_size}, got {_mnnvl_workspace.shape[0]}"
+        )
+
+        ws_bytes = decode_cp_a2a_workspace_size(_cp_size)
+        expected_elems = (ws_bytes + 7) // 8  # int64 elements
+        assert _mnnvl_workspace.shape[1] == expected_elems
+        assert _mnnvl_workspace.dtype == torch.int64
+
+        _comm.Barrier()
```
[the remark this answers is not in the document]
Comment (davidjpyu) [reply tests/comm/test_mnnvl_dcp_alltoall.py]: All existence of _comm.Barrier() calls are now changed to try/finally structure.

File: csrc/nv_internal/tensorrt_llm/kernels/helixAllToAll.cu [hunk 5]
```diff
@@ -0,0 +1,641 @@
...
+  static std::unordered_map<std::tuple<int, int, int>, std::tuple<int, int, int>, hash_cache_key>
+      cache;
+  int deviceId = 0;
+  TLLM_CUDA_CHECK(cudaGetDevice(&deviceId));
+  int singleShmSize = std::max(computeTotalUnpackedSize(fields), computeProtoTransferSize(fields));
+  auto key = std::make_tuple(deviceId, cpSize, singleShmSize);
+  auto it = cache.find(key);
+  if (it != cache.end()) {
+    return it->second;
+  }
```
[the remark this answers is not in the document]
Comment (davidjpyu) [reply csrc/nv_internal/tensorrt_llm/kernels/helixAllToAll.cu:495]: fixed

File: csrc/nv_internal/cpp/common/envUtils.cpp [hunk 6]
```diff
@@ -356,4 +356,23 @@ int getEnvMoeA2ACombineBlockSize() {
...
+      char const* env = std::getenv("TRTLLM_ENABLE_PDL");
+      if (env) {
+        if (env[0] == '1' && env[1] == '\0') {
+          enablePDL = true;
+        } else if (env[0] == '0' && env[1] == '\0') {
+          enablePDL = false;
+        }
+      };
+    }
+  });
```
[the remark this answers is not in the document]
Comment (davidjpyu) [reply csrc/nv_internal/cpp/common/envUtils.cpp]: Don't think lock is needed as host-side launch is single-threaded in practice.

**Review by aleozlx — APPROVED** [review]

**Comment by aleozlx** [comment]
possible AOT test regression on main from another PR 
the fix:
https://github.com/flashinfer-ai/flashinfer/pull/3056
